src/core/tsi/alts/handshaker/alts_handshaker_client.cc

```diff
@@ -94,45 +117,127 @@ static bool is_handshake_finished_properly(grpc_gcp_HandshakerResp* resp) {
   return false;
 }
 
+static void alts_grpc_handshaker_client_unref(
+    alts_grpc_handshaker_client* client) {
+  if (gpr_unref(&client->refs)) {
+    if (client->base.vtable != nullptr &&
+        client->base.vtable->destruct != nullptr) {
+      client->base.vtable->destruct(&client->base);
+    }
+    grpc_byte_buffer_destroy(client->send_buffer);
+    grpc_byte_buffer_destroy(client->recv_buffer);
+    client->send_buffer = nullptr;
+    client->recv_buffer = nullptr;
+    grpc_metadata_array_destroy(&client->recv_initial_metadata);
+    grpc_slice_unref_internal(client->recv_bytes);
+    grpc_slice_unref_internal(client->target_name);
+    grpc_alts_credentials_options_destroy(client->options);
+    gpr_free(client->buffer);
+    grpc_slice_unref_internal(client->handshake_status_details);
+    gpr_mu_destroy(&client->mu);
+    gpr_free(client);
+  }
+}
+
+static void maybe_complete_tsi_next(
+    alts_grpc_handshaker_client* client, bool receive_status_finished,
+    recv_message_result* pending_recv_message_result) {
+  recv_message_result* r;
+  {
+    grpc_core::MutexLock lock(&client->mu);
+    client->receive_status_finished |= receive_status_finished;
+    if (pending_recv_message_result != nullptr) {
+      GPR_ASSERT(client->pending_recv_message_result == nullptr);
+      client->pending_recv_message_result = pending_recv_message_result;
+    }
+    if (client->pending_recv_message_result == nullptr) {
+      return;
+    }
+    const bool have_final_result =
+        client->pending_recv_message_result->result != nullptr ||
+        client->pending_recv_message_result->status != TSI_OK;
+    if (have_final_result && !client->receive_status_finished) {
```

Comment:
Could you please add some comments about this line? e.g., it returns when 1) handshake successfully completes (`client->pending_recv_message_result->result != nullptr`) or 2) handshake fails (`client->pending_recv_message_result->status != TSI_OK`) or client has not received response on client's `RECV_STATUS_ON_CLIENT`.

Reply:
It seems that we will not run TSI next callback when 1) Handshake completes and succeeds (`client->pending_recv_message_result->result != nullptr`) or handshake fails (`client->pending_recv_message_result->status != TSI_OK`)  and 2) we do not receive the response on client's `RECV_STATUS_ON_CLIENT` op. Is it because if handshake server returns response for `RECV_MESSAGE`, it will not return for `RECV_STATUS_ON_CLIENT`?  Also, if the handshake fails, should not we also invoke the TSI next callback by feeding in corresponding error status?

Reply:
Please ignore my second question. I think your response to my questions below (i.e., calling TSI next callback twice) answers it :)

Reply:
> Could you please add some comments about this line? e.g., it returns when 1) handshake successfully completes (client->pending_recv_message_result->result != nullptr) or 2) handshake fails (client->pending_recv_message_result->status != TSI_OK) or client has not received response on client's RECV_STATUS_ON_CLIENT.

Yes, we return early here if "handshake successfully completes or handshake fails" and "not received status'". I added this comment

Reply:
> Also, if the handshake fails, should not we also invoke the TSI next callback by feeding in corresponding error status?

I do think it would make sense to override an otherwise-OK TSI next status with an error if the RECV_STATUS op indicates that the handshake RPC has failed. I've left a TODO for this though; if doing this, propose saving it for another PR.

Reply:
SGTM